src/java.base/share/classes/java/lang/Object.java

```diff
@@ -95,13 +114,26 @@ public Object() {}
      * As far as is reasonably practical, the {@code hashCode} method defined
      * by class {@code Object} returns distinct integers for distinct objects.
      *
+     * <p>This method should <em>not</em> throw any exceptions during
+     * typical operation. The implementation should not use excessive
+     * memory or time for its computations. 
+     *
+     * The {@code hashCode} method may be called frequently, possibly
+     * in the context of {@linkplain java.util##CollectionsFramework
+     * collections}.
+     *
+     * As as a quality of implementation concern, a particular
+     * implementation of this method may or may not support generating
```

Comment:
"may not support generating hash codes" sounds weird; maybe like "may or may not guard against cyclic data structures in recursive hash code generation."

I think the key is "recursive" here, as only recursive implementations are at risk of these faults.

Reply:
Well, recursion is the easiest way to accidentally write an infinite loop, but I think a few other ones would be possible too ;-) I'll considering updating the wording to highlight the most likely hazards; thanks.

Reply:
Unfortunately the typical implementation pattern is to recur through the component fields of the object. Not recursion through `this` but rather the hashCode implementation of some object will typically invoke hashCode of its component fields. Of course this can result in infinite recursion (really, StackOverflowError) if the object graph contains cycles, but in general, implementations of these methods don't guard against this possibility. (Nor should they. Adding such guard code seems impractical.)

Note that the default implementations of these methods for records and for Collection classes doesn't guard against cycles. (The collections contain some special cases for collection that contains itself, but not cycles in general.)

I have a general concern about advice that isn't actionable, such as "may not support" or "care should be taken" etc. If we really want to say something about cyclic data structures, maybe we could say something like this. For data structures that may contain cycles, these implementations should avoid traversing component fields that may lead to cycles. A class could simply inherit the method implementations from Object. Alternatively, the class could contain a unique ID that's used for determining the result of equals/hashCode/toString. (Of course, determining uniqueness is up to the application.)